lib/truffle/java/proxy_builder.rb

```diff
@@ -0,0 +1,326 @@
+#
+# Eclipse Public License version 1.0
+# GNU General Public License version 2
+# GNU Lesser General Public License version 2.1
+
+module JavaUtilities
+  class JavaProxyBuilder
+
+    def initialize(a_proxy, java_class)
+      @java_class = java_class
+      @proxy = a_proxy
+      @static_members = {}
+      @instance_members = {}
+      reserve_names()
+    end
+
+    def reserve_names()
+      both = ["__id__", "__send__", "instance_of?"]
+      instance = ["class", "initialize"]
+      static = ["new"]
+      (both + instance).map { |name| add_to_instance(name, Reserved.new) }
+      (both + static).map { |name| add_to_singleton(name, Reserved.new) }
+    end
+
+    def add_to_singleton(name, thing, maybe_getter=false, maybe_setter=false)
+      add_to_map(@static_members, name, thing, maybe_getter, maybe_setter)
+    end
+
+    def add_to_instance(name, thing, maybe_getter=false, maybe_setter=false)
+      add_to_map(@instance_members, name, thing, maybe_getter, maybe_setter)
+    end
+
+    def add_to_map(a_map, name, thing, maybe_getter, maybe_setter)
+      add_to_map_internal(a_map, name, thing)
+      alias_names(name, maybe_getter, maybe_setter) do |a|
+        add_to_map_internal(a_map, a, Alias.new(a, name))
+      end
+    end
+
+    def alias_names(name, maybe_getter, maybe_setter)
+      return if name.upcase == name
+      parts = name.split(/(?=[[:upper:]])/).map { |v| v.downcase }
+      return if parts.size == 1
+      getter = maybe_getter & (parts[0] == "get")
+      setter = maybe_setter & (parts[0] == "set")
+      yield parts.join("_")
+      yield parts[1..-1].join("_") + "?" if parts[0] == "is"
+      yield parts[1..-1].join("_") if getter
+      yield parts[1..-1].join("_") + "=" if setter
+    end
+
+    def add_to_map_internal(a_map, name, thing)
+      return a_map[name] = thing if ! a_map.has_key?(name)
+      return a_map[name] = thing if a_map[name].precedence > thing.precedence
+      return a_map[name].combine_with(thing) if a_map[name].precedence == thing.precedence
+      return a_map[name]
+    end
+
+    def add_interfaces
+      interfaces = Java.invoke_java_method(CLASS_GET_INTERFACES, @java_class)
+      # Not using idiomatic Ruby here as we might not have bootstrapped that at this point.
+      interfaces_size = JavaUtilities.java_array_size(interfaces)
+      (0...interfaces_size).each do |i|
+        interface_proxy = JavaUtilities.make_proxy(JavaUtilities.java_array_get(interfaces, i))
+        @proxy.include(interface_proxy) unless @proxy.ancestors.include?(interface_proxy)
+      end
+      self
+    end
+
+    def add_static_fields
+      fields = Java.invoke_java_method(CLASS_GET_DECLARED_FIELDS, @java_class)
+      # Not using idiomatic Ruby here as we might not have bootstrapped that at this point.
+      fields_size = JavaUtilities.java_array_size(fields)
+      (0...fields_size).each do |i; f, mh, name, getter, setter, is_static, is_const, const_val|
```

Comment:
Why are these parameters declared as block parameters?
I guess we don't need that.
Does `bootstrapped` means JRuby boostrapping?